crates/matrix-sdk-crypto/src/olm/account.rs

```diff
@@ -543,17 +556,55 @@ impl Account {
         Some(key_count as u64)
     }
 
-    pub(crate) fn generate_fallback_key_helper(&mut self) {
-        if self.inner.fallback_key().is_empty() {
+    pub(crate) fn generate_fallback_key_if_needed(&mut self) {
+        if self.inner.fallback_key().is_empty() && self.fallback_key_expired() {
             let removed_fallback_key = self.inner.generate_fallback_key();
+            self.fallback_creation_timestamp = Some(MilliSecondsSinceUnixEpoch::now());
 
             debug!(
                 ?removed_fallback_key,
-                "No unused fallback keys were found on the server, generated a new fallback key.",
+                "The fallback key either expired or we didn't have one: generated a new fallback key.",
             );
         }
     }
 
+    /// Check if our most recent fallback key has expired.
+    ///
+    /// We consider the fallback key to be expired if it's older than a week.
+    /// This is the lower bound for the recommended signed pre-key bundle
+    /// rotation interval in the X3DH spec[1].
+    ///
+    /// [1]: https://signal.org/docs/specifications/x3dh/#publishing-keys
+    fn fallback_key_expired(&self) -> bool {
+        const FALLBACK_KEY_MAX_AGE: Duration = Duration::from_secs(3600 * 24 * 7);
```

Comment:
Please can this be configurable by the caller, as for complement-crypto tests I'd want to set this to be a few seconds to ensure keys are being cycled, which means it needs to be visible to the FFI bindings.

Reply:
I might be able to use something like https://github.com/wolfcw/libfaketime but it'll be very brittle and likely to break randomly.

A compile-time feature flag works for me, as that can then be used for WASM bindings as well.

Reply:
There are other places where this would be nice e.g we cannot test `rotation_period_ms` because rust SDK has a lower bound of 1 hour.

Reply:
+1 for allowing to provide config either at runtime or compile time - the integration tests we are writing use real homeservers in separate processes, so faking out the time at this level is difficult.

It's also arguable, at least in the case of `rotation_period_ms` that different use-cases might want different limits on these values, so config would be helpful for non-test situations. (E.g. a [password manager](https://github.com/matrix-org/matrix-spec-proposals/pull/4114) might have different security requirements from a chat app).

Reply:
You can set the `rotation_period_ms` with the `m.room.encryption` event, it's just that we clamp the values so you don't get stuck in a loop where by the time you share the room key it already expired.

Reply:
Yeah, we are setting the `rotation_period_ms` with the `m.room.encryption` event but running into the clamp because we want our test to run in <1hr